OpenRA.Mods.Common/Traits/DeliversExperience.cs

```diff
@@ -0,0 +1,133 @@
+	class DeliversExperienceInfo : ITraitInfo, Requires<GainsExperienceInfo>
+	{
+		[Desc("The amount of experience the donating player receives.")]
+		public readonly int PlayerExperience = 0;
+
+		[Desc("Identifier checked against AcceptsDeliveredExperience.ValidTypes. Only needed if the latter is not empty.")]
+		public readonly string Type = null;
+
+		[VoiceReference] public readonly string Voice = "Action";
+
+		public object Create(ActorInitializer init) { return new DeliversExperience(init, this); }
+	}
+
+	class DeliversExperience : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly DeliversExperienceInfo info;
+		readonly Actor self;
+
+		public DeliversExperience(ActorInitializer init, DeliversExperienceInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				if (self.Trait<GainsExperience>().Level != 0)
+					yield return new DeliversExperienceOrderTargeter();
+			}
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (target.Actor.Trait<GainsExperience>().Level == target.Actor.Trait<GainsExperience>().MaxLevel)
+				return null;
+
+			if (order.OrderID != "DeliverExperience")
+				return null;
+
+			if (target.Type == TargetType.FrozenActor)
+				return new Order(order.OrderID, self, queued) { ExtraData = target.FrozenActor.ID };
+
+			return new Order(order.OrderID, self, queued) { TargetActor = target.Actor };
+		}
+
+		public string VoicePhraseForOrder(Actor self, Order order)
+		{
+			return info.Voice;
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString != "DeliverExperience")
+				return;
+
+			var target = self.ResolveFrozenActorOrder(order, Color.Yellow);
+			if (target.Type != TargetType.Actor)
+				return;
+
+			if (target.Actor.Trait<GainsExperience>().Level == target.Actor.Trait<GainsExperience>().MaxLevel)
+				return;
+
+			if (!order.Queued)
+				self.CancelActivity();
+
+			var level = self.Trait<GainsExperience>().Level;
+
+			self.SetTargetLine(target, Color.Yellow);
+			self.QueueActivity(new DonateExperience(self, target.Actor, level, info.PlayerExperience));
+		}
+
+		public class DeliversExperienceOrderTargeter : UnitOrderTargeter
+		{
+			public DeliversExperienceOrderTargeter()
+				: base("DeliverExperience", 5, "enter", false, true) { }
+
+			public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
+			{
+				var type = self.Info.TraitInfo<DeliversExperienceInfo>().Type;
+				var targetInfo = target.Info.TraitInfoOrDefault<AcceptsDeliveredExperienceInfo>();
+				var targetGainsExperience = target.TraitOrDefault<GainsExperience>();
+
+				if (targetGainsExperience == null || targetInfo == null)
+					return false;
+
+				if (targetGainsExperience.Level == targetGainsExperience.MaxLevel)
+					return false;
+
+				return targetInfo != null
```

Comment:
You're already checking `targetInfo == null` above, so this shouldn't be needed.